lib/streaming-pipeline.js

```diff
@@ -0,0 +1,193 @@
+const {Writable} = require('stream');
+const validateArgumentTransformers = require('./argument-transformer-validator');
+const RiffError = require('./riff-error');
+const OutputMarshaller = require('./output-marshaller');
+const logger = require('util').debuglog('riff');
+const InputUnmarshaller = require('./input-unmarshaller');
+const {guardWithTimeout} = require('./async');
+
+const toObject = (parameters) => {
+    const result = {};
+    for (let i = 0; i < parameters.length; i++) {
+        result[`${i}`] = parameters[i];
+    }
+    return result;
+};
+
+const validateArity = (arity) => {
+    if (typeof arity === 'undefined') {
+        throw new RiffError('error-function-arity', 'Cannot determine function arity. Aborting now');
+    }
+    if (typeof arity !== 'number' || Math.floor(arity) !== arity || arity <= 0) {
+        throw new RiffError('error-function-arity', `Function arity must be an integer >= 1, received: ${arity}. Aborting now`);
+    }
+};
+
+const validateTransformers = (transformers) => {
+    if (typeof transformers === 'undefined') {
+        return;
+    }
+    validateArgumentTransformers(transformers);
+};
+
+module.exports = class StreamingPipeline extends Writable {
+
+    constructor(userFunction, destinationStream, options) {
+        super(options);
+        const arity = userFunction['$arity'];
+        validateArity(arity);
+        validateTransformers(userFunction['$argumentTransformers']);
+        this._options = options;
+        this._hookTimeoutInMs = options['hookTimeoutInMs'] || 10000;
+        this._userFunction = userFunction;
+        this._destinationStream = destinationStream;
+        this._startReceived = false;
+        this._functionInputs = [];
+        this._functionOutputs = [];
+        this._finishedOutputs = 0;
+        this.on('finish', () => {
+            logger('Ending input streams');
+            this._endInputs();
+        });
+        this.on('error', () => {
+            logger('Error occurred - ending pipeline now');
+            this.end();
+        });
+        this._invokeHooks();
+        logger('Streaming pipeline initialized');
+    }
+
+    _invokeHooks() {
+        const hookResolve = (hookName) => {
+            return (timerId) => {
+                clearTimeout(timerId);
```

Comment:
timeouts should be set and cleared in the same module, ideally in the same function to avoid the responsibility leaking out.